lib/github/tests/test_api.nit

```diff
@@ -0,0 +1,344 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module test_api is test
+
+intrude import api
+
+# GithubAPI testing
+#
+# To avoid test flakyness we test the GithubAPI against a mock of the real one.
+# For each api request we return a cache file of the real API response body.
+#
+# Cache files can be automatically created and updated by setting
+# `update_responses_cache` to `true` then running `nitunit`.
+class MockGihubCurl
+	super GithubCurl
+
+	# Mock so it returns the response from a file
+	#
+	# See `update_responses_cache`.
+	redef fun get_and_parse(uri) do
+		print uri # for debugging
+
+		var path = uri.replace("https://api.github.com/", "/")
+		assert has_response(path)
+
+		if update_responses_cache then
+			var file = response_file(path)
+			save_actual_response(uri, file)
+		end
+
+		var response = response_string(path).parse_json
+		if response_is_error(path) then
+			var title = "GithubAPIError"
+			var msg = response.as(JsonObject)["message"].as(String)
+			var err = new GithubError(msg, title)
+			err.json["requested_uri"] = uri
+			err.json["status_code"] = response_code(path)
+			return err
+		end
+		return response
+	end
+
+	var test_responses: Map[String, String] do
+		var map = new HashMap[String, String]
+		map["/user"] = "user_Morriar"
+		map["/users/Morriar"] = "user_Morriar"
+		map["/repos/nitlang/nit"] = "repo_nit"
+		map["/repos/nitlang/nit/labels/ok_will_merge"] = "repo_labels_ok_will_merge"
+		map["/repos/nitlang/nit/milestones/4"] = "repo_milestones_4"
+		map["/repos/nitlang/nit/branches"] = "repo_branches_nit"
+		map["/repos/nitlang/nit/branches/master"] = "repo_branches_master"
+		map["/repos/nitlang/nit/issues/1000"] = "repo_issues_1000"
+		map["/repos/nitlang/nit/issues/comments/6020149"] = "repo_issues_comments_6020149"
+		map["/repos/nitlang/nit/issues/events/199674194"] = "repo_issues_events_199674194"
+		map["/repos/nitlang/nit/pulls/1000"] = "repo_pulls_1000"
+		map["/repos/nitlang/nit/commits/64ce1f"] = "repo_commits_64ce1f"
+		map["/repos/nitlang/nit/comments/8982707"] = "repo_comments_8982707"
+		map["/repos/nitlang/nit/pulls/comments/21010363"] = "repo_pulls_comment_21010363"
```

Comment:
Did that originally then I changed it back for two reasons:

1. To define the set of possible callable URIs, so if there is a call outside that the test will fail
2. To define the set of possible readable files to avoid loading arbitrary JSON in case some freaking coincidence could expose some secrets in our CI or I don't know.

Instead of defining both arrays I went crazy and defined the map.